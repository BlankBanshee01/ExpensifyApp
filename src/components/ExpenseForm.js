import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";

// const date = new Date();
const now = moment();

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide a description and an amount for your expense",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  }
}

// import React, { Component } from "react";
// import moment from "moment";
// import { SingleDatePicker } from "react-dates";
// import "react-dates/lib/css/_datepicker.css";
// // import "react-dates/initialize";

// export default class ExpenseForm extends Component {
//   state = {
//     description: "",
//     amount: "",
//     note: "",
//     createdAt: moment(),
//     focused: false,
//   };

//   handleDescription = (e) => {
//     const description = e.target.value;
//     this.setState(() => ({ description }));
//   };

//   handleAmount = (e) => {
//     const amount = e.target.value;
//     if (amount.match(/^\d*(\.\d{0,2})?$/)) {
//       this.setState(() => ({ amount }));
//     }
//   };

//   handleNote = (e) => {
//     const note = e.target.value;
//     this.setState(() => ({ note }));
//   };

//   render() {
//     return (
//       <div>
//         <form>
//           <input
//             type="text"
//             placeholder="Description"
//             autoFocus
//             value={this.state.description}
//             onChange={this.handleDescription}
//           />
//           <SingleDatePicker
//             date={this.state.date} // momentPropTypes.momentObj or null
//             onDateChange={(createdAt) => this.setState({ createdAt })} // PropTypes.func.isRequired
//             focused={this.state.focused} // PropTypes.bool
//             onFocusChange={({ focused }) => this.setState({ focused })} // PropTypes.func.isRequired
//             // id="your_unique_id" // PropTypes.string.isRequired,
//           />
//           <input
//             type="text"
//             placeholder="Amount"
//             value={this.state.amount}
//             onChange={this.handleAmount}
//           />
//           <textarea
//             placeholder="Add a note (optional)"
//             value={this.state.note}
//             onChange={this.handleNote}
//           />
//         </form>
//       </div>
//     );
//   }
// }
