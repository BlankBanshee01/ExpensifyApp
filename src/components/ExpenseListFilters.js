import React, { Component } from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";

export class ExpenseListFilters extends Component {
  state = {
    focusedInput: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (focusedInput) => this.setState({ focusedInput });

  onTextFilterChange = (e) => this.props.setTextFilter(e.target.value);

  onSortByFilterChange = (e) => {
    if (e.target.value === "amount") {
      this.props.sortByAmount();
    } else if (e.target.value === "date") {
      this.props.sortByDate();
    }
  };
  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextFilterChange}
        ></input>
        <select
          value={this.props.filters.sortBy}
          onChange={this.onSortByFilterChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.focusedInput}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  setTextFilter: (value) => dispatch(setTextFilter(value)),
  sortByAmount: () => dispatch(sortByAmount()),
  sortByDate: () => dispatch(sortByDate()),
});

const mapStateToProps = (state) => ({
  filters: state.filters,
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
