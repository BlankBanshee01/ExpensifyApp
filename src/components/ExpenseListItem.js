import React from "react";
import { Link } from "react-router-dom";

const expenseListItem = ({ description, amount, createdAt, id }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>Amount: {amount}</p>
      <p>Date: {createdAt}</p>
    </div>
  );
};

export default expenseListItem;
