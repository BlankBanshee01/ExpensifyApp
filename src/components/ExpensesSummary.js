import React from "react";
import { connect } from "react-redux";
import numeral from "numeral";
import { Link } from "react-router-dom";
import selectExpenseTotal from "../selectors/selectExpenseTotal";
import selectExpenses from "../selectors/expenses";

export const ExpensesSummary = (props) => {
  const expenseWord = props.visibleExpenseNumber === 1 ? "expense" : "expenses";
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Showing a total of <span>{props.visibleExpenseNumber}</span>{" "}
          {expenseWord} totaling{" "}
          <span>{numeral(props.totalAmount / 100).format("$0,0.00")}</span>
        </h1>
        <div className="page-header__actions">
          <Link to="/add_expense" className="button">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    visibleExpenseNumber: visibleExpenses.length,
    totalAmount: selectExpenseTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);
