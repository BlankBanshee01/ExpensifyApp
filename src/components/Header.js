import React from "react";
import { NavLink } from "react-router-dom";

const Header = () => (
  <div>
    <h1>Expensify</h1>
    <NavLink activeClassName="active-link" exact to="/">
      Home
    </NavLink>
    <NavLink activeClassName="active-link" to="/add_expense">
      Create Expence
    </NavLink>
    <NavLink activeClassName="active-link" to="/help">
      Help
    </NavLink>
  </div>
);

export default Header;
