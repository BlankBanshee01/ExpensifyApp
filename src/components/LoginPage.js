import React from "react";
import { startLogin } from "../actions/auth";
import { connect } from "react-redux";

export const LoginPage = (props) => (
  <div className="box-layout">
    <div className="box-layout__box">
      <h1 className="box-layout__title">Expensify</h1>
      <p>Control your expenses</p>
      <button onClick={props.startLogin} className="button">
        Login with Google
      </button>
    </div>
  </div>
);

const mapDispatchToProps = (dispatch) => ({
  startLogin: () => dispatch(startLogin()),
});

export default connect(undefined, mapDispatchToProps)(LoginPage);
