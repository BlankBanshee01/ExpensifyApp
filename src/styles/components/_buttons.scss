.button {
  background-color: $blue;
  border: 0;
  color: white;
  display: inline-block;
  font-size: $font-size-l;
  font-weight: 300;
  line-height: 1;
  padding: $m-size;
  text-decoration: none;
}

.button--link {
  background: none;
}

.button--secondary {
  background: #888888;
}
