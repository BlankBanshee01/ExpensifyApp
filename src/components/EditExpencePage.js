import React, { Component } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends Component {
  onSubmit = (updatedExpense) => {
    this.props.editExpense(this.props.expense.id, updatedExpense);
    this.props.history.push("/");
  };

  onClick = () => {
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
        <button onClick={this.onClick}>Remove Expense</button>
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  editExpense: (id, updatedExpense) =>
    dispatch(editExpense(id, updatedExpense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
});

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(
    (expense) => expense.id === props.match.params.id
  ),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
