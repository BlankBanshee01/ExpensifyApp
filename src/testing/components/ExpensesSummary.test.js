import React from "react";
import { shallow } from "enzyme";
import { ExpensesSummary } from "../../components/ExpensesSummary";

test("should render expense summary with 1 expense", () => {
  const wrapper = shallow(
    <ExpensesSummary visibleExpenseNumber={1} totalAmount={20} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render expense summary with multiple expense {S}", () => {
  const wrapper = shallow(
    <ExpensesSummary visibleExpenseNumber={3} totalAmount={20} />
  );
  expect(wrapper).toMatchSnapshot();
});
