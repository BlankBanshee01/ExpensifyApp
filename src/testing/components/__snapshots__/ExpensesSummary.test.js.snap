// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Showing a total of 
      <span>
        1
      </span>
       
      expense
       totaling
       
      <span>
        $0.20
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/add_expense"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render expense summary with multiple expense {S} 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Showing a total of 
      <span>
        3
      </span>
       
      expenses
       totaling
       
      <span>
        $0.20
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/add_expense"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
